pkg/serdes.go

```diff
@@ -37,83 +37,112 @@ func readBodyBuffered(resp *http.Response) ([]byte, error) {
-	originalWidth, originalHeight := img.Width(), img.Height()
-
-	if transform != nil {
-		aspectRatio, err = transform.cropAndResizeToClosestAspectRatio(img, aspectRatio)
-		if err != nil {
-			return nil, -1., err
-		}
-	}
-
-	width, height := img.Width(), img.Height()
-
 	// If the image is 4 channels, we need to drop the alpha channel
 	if img.Bands() == 4 {
 		err = img.Flatten(&vips.Color{R: 255, G: 255, B: 255}) // Flatten with white background
 		if err != nil {
-			fmt.Println("Error flattening image:", err)
-			return nil, -1., err
+			return nil, -1., fmt.Errorf("error flattening image: %w", err)
 		}
 		fmt.Println("Image flattened")
 	}
 
 	// If the image is not a mask but is 1 channel, we want to convert it to 3 channels
 	if (img.Bands() == 1) && !isMask {
-		err = img.ToColorSpace(vips.InterpretationSRGB)
+		if img.Metadata().Format == vips.ImageTypeJPEG {
+			err = img.ToColorSpace(vips.InterpretationSRGB)
+			if err != nil {
+				return nil, -1., fmt.Errorf("error converting to sRGB: %w", err)
+			}
+		} else {
+			// // FIXME: maybe that we could recover these still. By default throws an error, sRGB and PNG not supported
+			return nil, -1., fmt.Errorf("1 channel PNG image not supported")
+		}
+	}
+
+	// If the image is 2 channels, that's gray+alpha and we flatten it
+	if img.Bands() == 2 {
+		err = img.ExtractBand(1, 1)
+		fmt.Println("Gray+alpha image, removing alpha")
 		if err != nil {
-			fmt.Println("Error converting to sRGB:", err)
-			return nil, -1., err
+			return nil, -1., fmt.Errorf("error extracting band: %w", err)
 		}
 	}
 
+	// Optionally crop and resize the image on the fly. Save the aspect ratio in the process for future use
+	originalWidth, originalHeight := img.Width(), img.Height()
+
+	if transform != nil {
+		// Catch possible SIGABRT in libvips and recover from it
+		defer func() {
+			if r := recover(); r != nil {
+				if strings.Contains(fmt.Sprint(r), "SIGABRT") {
+					err = fmt.Errorf("caught SIGABRT: %v", r)
+				} else {
+					panic(r) // re-throw the panic if it's not SIGABRT
+				}
+			}
+		}()
+
+		aspectRatio, err = transform.cropAndResizeToClosestAspectRatio(img, aspectRatio)
+		if err != nil {
+			return nil, -1., fmt.Errorf("error cropping and resizing image: %w", err)
+		}
+	}
+
+	width, height := img.Width(), img.Height()
+
 	// If requested, re-encode the image to a jpg or png
 	var imgBytes []byte
 	var channels int
 	var bitDepth int
 
 	if encodeImage {
-		if err != nil {
-			return nil, -1., err
-		}
-
 		if img.Bands() == 3 {
 			// Re-encode the image to a jpg
 			imgBytes, _, err = img.ExportJpeg(&vips.JpegExportParams{Quality: 95})
-			if err != nil {
-				return nil, -1., err
-			}
 		} else {
 			// Re-encode the image to a png
-			imgBytes, _, err = img.ExportPng(vips.NewPngExportParams())
-			if err != nil {
-				return nil, -1., err
-			}
+			imgBytes, _, err = img.ExportPng(&vips.PngExportParams{
```

Comment:
same as in the loading case, be explicit above the params used here instead of using ones which would depend on the library defaults